Object curly spacing

jmlopez-rod opened the pull request:
This PR adds the `object-curly-spacing` rules according to specifications from [eslint](http://eslint.org/docs/rules/object-curly-spacing.html). The rule and test files have have been added and the README file has been updated to show that it is no longer in the TODO list.

Several other changes have been made in order to improve the developing experience.

- **No longer running all the tests to check if the rule in development works**:

  By default `gulp test` will run all the tests that it can find. We can now target a particular
  test by using the `single` argument as explained in the README: `gulp test --single rule-name`.

- **Transpile files even if linting fails**:

  When running the tests we always run the build task which depends on the linting task.
  Sometimes we may encounter linting errors that we are not willing to clean up yet and this
  should not stop us from writing the rule. When this is the case we can add the flag `--no-lint`
  so that the linting task is not triggered. For instance `gulp test --single rule-name --no-lint`.

I also added the rule to the project to test if we have inconsistencies in the project files. So far we are not adding spaces on imports but we do on object literals. In order to leave things working I have turned off the rule but we should consider using it in order to avoid future inconsistencies. I will open an issue for this.

Finally, the current state of the project is bad. Please test your master branch and let me know if you the `valid-jsdoc` test fails at `should fail when using invalid JSDoc comments (matchDescription: "regex", requireReturn: false)`.

jmlopez-rod commented:
closes #58

jmlopez-rod commented on src/rules/objectCurlySpacingRule.ts:
```diff
@@ -0,0 +1,71 @@
+  }
+}
+
+class BlockSpacingWalker extends Lint.RuleWalker {
```
Need to change the name. This was a copy/paste mistake from the BlockSpacing rule.

jmlopez-rod commented:
We still need to merge PR #100 in order for the continuous integration to pass. In the mean time, once this is merged and my `no-multi-spaces` branch is approved then we should consider adding these tests as well:

https://github.com/eslint/eslint/blob/master/tests/lib/rules/object-curly-spacing.js

Hopefully none will fail.

buzinas commented:
Please, undo `.gitignore`, `gulpfile.js` and `package.json`.

There is a reason for not transpiling files when tests fail, and this is something I won't change. And there is also a reason for running all tests any time you want to test something.

Both "improvements" are design choices, and not errors.

jmlopez-rod commented:
Never said they were errors, they are just a way to speed up the development process for the rule we are focusing on. As the number of rules and tests grow more the time it takes to run over all the test will continue to grow. Those changes do not interfere with the current setup in any way. As the final step the developer needs to be responsible to run the final `gulp test` to make sure all the tests are running before pushing the changes so that the continous-integration won't fail.

Please note that the `--no-lint` flag is so that the traspilation won't fail due to lint errors (as opposed to tests failing) which can happen quite often (max-line, no-var-used, etc). During development I'm aware that some of these errors are introduced often but this shouldn't stop the tests from running, especially when I'm not ready to commit and create a PR.

The only reason I can think of we need to run all the tests when we want to test a single rule is if some of the test share some of the rules and to see how they interact with each other. Which is why when I was done writing my rule I ran the `gulp test` to make sure everything was in order.
It was because of the `gulp test` that I found out the one of your tests was broken. If I didn't write this `flag` I would have just given up and not written this PR and the other stuff I have written.

Don't get me wrong, I love the work you have done, it was because of it that it inspired me to stop waiting for the rules to be done and instead start writing them so that others may use them.

I ask you to please reconsider your views on those two optional flags I wrote. If nothing has changed your mind please let me know if I should I also revert the change to the `tslint.config` file. Thank you for your time.

jmlopez-rod commented:
Update: I have set the branch up to date with master and while testing I noticed that the `tslint.config` cannot have the rules of the project applied to it since this breaks the `gulp test` flow. Notice that now all the tests are passing and how these flags did not interfere.

Also, on the `.gitignore`, do you really want to commit js files on the src directory?

buzinas commented:
CI is failing because of the new `"rulesDirectory"` property in `tslint.json`.

About what you said, OK, I reconsider, and we can keep your changes.

About your last question: we should never have js files in the src directory, so I don't understand why we should have this restriction.

jmlopez-rod commented:
Yeah, I deleted my `dist` folder and i got the error that `CI` encountered, that's why I removed the changes from it. Any way we can apply the rules we make here to the project without interrupting the linter? I think that'll be for another issue.

Thank for you reconsidering. I have the next PR ready to go.

I only brought the last point since you had mentioned to revert the .gitignore. Either way, I'm glad you agreed that we should not have them. That is all. Thanks.

buzinas commented:
Thanks for contributing, have a good week :)
